### DISCUSSION TITLE
Better package version resolution accross meteor releases

### DISCUSSION THREAD

**Discussion by vparpoil** [meteor/meteor#13142] [Feature Requests]
Package version resolution is at the moment tied to the specific release of Meteor.
Indeed, the 3.0-rc.1 doesn't point to the same version of packages than beta.6 or alpha.1

Here is the output I have when trying to upgrade to the last RC :
```shell
Conflict: Constraint mongo@1.10.1 || 1.12.0 is not satisfied by mongo 2.0.0-rc300.2.
Constraints on package "mongo":
* mongo@2.0.0-rc300.2 <- top level
* mongo@~2.0.0-rc300.2 <- top level
* mongo@1.10.1 || 1.12.0 || 2.0.0-beta300.0 <- aldeed:collection2 4.0.0
* mongo@1.12.0 || 1.16.1 || 2.0.0-beta300.0 <- matb33:collection-hooks 1.3.1
* mongo@1.0.0||2.0.0||2.0.0-alpha300.17 <- quave:migrations 2.0.2
* mongo@1.16.8 || 2.0.0-rc300.0 <- ostrio:files 3.0.0-beta.6
* mongo@1.10.1 <- herteby:denormalize 0.6.7
* mongo@2.0.0-rc300.2 <- accounts-base 3.0.0-rc300.1 <- accounts-password 3.0.0-rc300.1
* mongo@2.0.0-alpha300.18 <- montiapm:agent 3.0.0-beta.4
* mongo@1.1.24||2.0.0-alpha300.10 <- montiapm:meteorx 2.3.1 <- montiapm:agent 3.0.0-beta.4
* mongo@1.10.1 || 1.12.0 <- jkuester:migration-helper 1.0.2
```

Several package versions are conflicting, BUT most of them have already been migrated to be ready for meteor 3, since you can see references to `-alpha300.18` or other `-beta300` and `-rc300`. However, at each bump of Meteor version, all this falls appart and we need to have a re-publish of all those packages or to clone locally to make the changes...

This creates a bad DX and a lot of friction for someone willing to migrate to the RC and provide feedback about it. Also it makes it almost impossible to start a new project based on the betas or RC, because you need to clone and maintain dozens of packages to follow up on meteor releases.
I think this feeling is shared by others ([see here the maintainer of universe:i18n explaining why they won’t publish a meteor-3 compatible version yet](https://github.com/vazco/meteor-universe-i18n/pull/188#issuecomment-2014824498))

It would be great if the future package version resolution could be more flexible in order to simplify this.

**Comment by perbergland** [comment]
I propose a "meteor-resolutions" section in package.json with the same syntax as for [yarn’s resolutions](https://yarnpkg.com/configuration/manifest#resolutions) where you can override package resolution on a global level.

That solves not only this scenario but also other scenarios where the application developers while understanding the risks involved can override package sub-dependencies.

The example here would be

```json
"meteor-resolutions": {
  "mongo": "mongo@2.0.0-rc300.2"
}
```

All transitive package dependencies on mongo would then resolve to mongo@2.0.0-rc300.2.

**Reply by zodern** [reply]
Meteor already supports this by adding a `!` after the package version in `.meteor/packages` to force Meteor to use a specific version. For example, if you use packages that depend on coffeescript@1, you can force Meteor to use coffeescript@2.

The problem with the `!` is it forces meteor use an exact version, which can make upgrading Meteor difficult. For example, if one version of the package is compatible with the current Meteor release, and a different version of the package is compatible with the new Meteor release, there is no way to tell Meteor it can use either versions.

One possible solution is to add a different indicator that is easier to work with. For example, adding `>!` after a package version forces Meteor to use at least the minimum version specified or any newer version.

**Comment by zodern** [comment]
In the example logs, there's only two packages that are causing the conflict: `jkuester:migration-helepr` and `hereteby:denormalize`. All of the constraints from the other packages are satisfied with `2.0.0-rc300.2`.

Packages that have already been updated to use a pre-release version of mongo@2 shouldn't need that constraint to be updated until there is a mongo@3. For example, `montiapm:meteorx` was published for alpha 10, and it is still compatible with the second release candidate.

I think there's three main problems:

1. Meteor's logs when there is a version constraint are confusing, especially if you don't have a good understanding of how it works. It could be significantly improved, for example by replacing `top level` with a clearer description and hiding by default the constraints on core packages that are compatible with the current release. Maybe there should also be a line to give more context on what is happening, something like `Unable to find version of ${package} that satisfies all constraints on it.`
2. There isn't much official documentation on this so it is difficult to accurately learn or understand how it works (there's a surprising amount of wrong information around).
3. It is possible to force Meteor to use a specific package version despite the constraints, but it isn't documented as far as I can tell and doesn't work well when upgrading Meteor. I shared more details [here](https://github.com/meteor/meteor/discussions/13142#discussioncomment-9439200)

I think it is important for the first or second item to be solved before the official Meteor 3 release to make it easier for people to upgrade.

**Comment by nachocodoner** [comment]
We have taken a note to consider the updates @zodern has suggested on this topic to promote a broader understanding of package version resolution and documentation, these is definitely worth for everyone.

**Comment by rj-david** [comment]
@nachocodoner, good to know that you (team) are working on this.

While on this topic, aside from making meteor 3 upgrade easier with less package conflicts, one big issue is the error message itself. The message is not developer-friendly. The forum is littered with support posts from developers asking for help with this error (I might have posted one or two of those when I was new to Meteor). The main reason was not about the constraint, but was due to developers who cannot understand what the error is about in the first place.

Maybe a simple summary of the error at the top can help:

The following packages are not supported due to version constraints:
- package A
- package B

[Existing constraint error here]
